File: DSA/SakshamDwivedi_590013147/Day24/D24P1.c
#include <stdio.h>
#include <stdlib.h>

struct node {
    int data;
    struct node* next;
};

int size = 0;
struct node* head = NULL;
void createNode(int);
void Print();
void Reverse();

int main(){
    createNode(1);
    createNode(2);
    createNode(3);
    createNode(4);
    createNode(5);
    createNode(6);
    printf("Before : ");
    Print();
    printf("\n");
    printf("After : ");
    Reverse();
    Print();
}

void createNode(int data){
       struct node* Node = (struct node*)malloc(sizeof(struct node));
       Node -> data = data;
       Node -> next = NULL;
       if(head == NULL){
            head = Node;
            size++;
            return;
       }
     struct node* current = head;
     while(current -> next != NULL){
      current = current -> next;
     }
     current -> next = Node;
     size++;
}

void Print(){
    struct node* current = head;
    while(current != NULL){
        if(current -> next != NULL){
        printf("%d -> ", current -> data);}
        else{printf("%d", current -> data);}
        current = current -> next;
    }

}

void Reverse() {
    struct node* current = head;
    struct node* ahead = head;
    struct node* before = NULL;
    while(current != NULL){
        ahead = ahead -> next;
        current -> next = before;
        before = current;
        current = ahead;
    }
    head = before;
}

File: DSA/SakshamDwivedi_590013147/Day24/D24P1.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

size = 0
head = None

def create_node(data):
    global head, size
    new_node = Node(data)
    if head is None:
        head = new_node
        size += 1
        return
    current = head
    while current.next:
        current = current.next
    current.next = new_node
    size += 1

def print_list():
    current = head
    while current:
        if current.next:
            print(current.data, " -> ", end="")
        else:
            print(current.data)
        current = current.next

def reverse_list():
    global head
    current = head
    ahead = head
    before = None
    while current:
        ahead = ahead.next
        current.next = before
        before = current
        current = ahead
    head = before

create_node(1)
create_node(2)
create_node(3)
create_node(4)
create_node(5)
create_node(6)

print("Before : ", end="")
print_list()

print("After : ", end="")
reverse_list()
print_list()
